py/objtype.c

```diff
@@ -85,14 +85,14 @@ static int instance_count_native_bases(const mp_obj_type_t *type, const mp_obj_t
 
 // This wrapper function allows a subclass of a native type to call the
 // __init__() method (corresponding to type->make_new) of the native type.
-static mp_obj_t native_base_init_wrapper(size_t n_args, const mp_obj_t *args) {
+static mp_obj_t native_base_init_wrapper(size_t n_args, const mp_obj_t *args, mp_map_t *kw_args) {
     mp_obj_instance_t *self = MP_OBJ_TO_PTR(args[0]);
     const mp_obj_type_t *native_base = NULL;
     instance_count_native_bases(self->base.type, &native_base);
-    self->subobj[0] = MP_OBJ_TYPE_GET_SLOT(native_base, make_new)(native_base, n_args - 1, 0, args + 1);
+    self->subobj[0] = MP_OBJ_TYPE_GET_SLOT(native_base, make_new)(native_base, n_args - 1, kw_args->used, args + 1);
```

Comment:
I'm not sure how this works: when calling `make_new`, the keyword arguments must be passed in the same `args` array as the positional arguments.  But upon entry to this function the positional in `args` are separate to the keyword in `kw_args`.

Is it implicitly relying on the fact that the keyword args do actually sit in the `args` array, but only by accident?

Reply:
> Is it implicitly relying on the fact that the keyword args do actually sit in the args array, but only by accident?

It's not by accident, the convention seems to be such that this is always the case. But there are 2 different ways in which this happens which is why my original comment about some things not being needed wasn't accurate because I only tested one case (I removed that part of the comment plus added extra tests to cover the case where `__init__` is not called with start arguments).

The part which this commit chages is indicated with `// ****`, rest is copy-paste of the actual code.

- calling `super().__init__()` or `super().__init__(pos0)` or `super().__init__(kw0=0)` or `super().__init__(pos0, kw0)` i.e. no star arguments, all result in dispatching `MP_BC_CALL_METHOD` which leads to the following call chain:

```
ENTRY(MP_BC_CALL_METHOD)
....

// args contains: fun  self/NULL  arg(0)  ...  arg(n_args-2)  arg(n_args-1)  kw_key(0)  kw_val(0)  ... kw_key(n_kw-1)  kw_val(n_kw-1)
mp_obj_t mp_call_method_n_kw(size_t n_args, size_t n_kw, const mp_obj_t *args);

// args contains, eg: arg0  arg1  key0  value0  key1  value1
mp_obj_t mp_call_function_n_kw(mp_obj_t fun_in, size_t n_args, size_t n_kw, const mp_obj_t *args) {
    // get the type
    const mp_obj_type_t *type = mp_obj_get_type(fun_in);
    // do the call
    if (MP_OBJ_TYPE_HAS_SLOT(type, call)) {
        return MP_OBJ_TYPE_GET_SLOT(type, call)(fun_in, n_args, n_kw, args);
    }
}

mp_obj_t fun_builtin_var_call(mp_obj_t self_in, size_t n_args, size_t n_kw, const mp_obj_t *args) {
    ...
    if (self->sig & 1) {
        // function allows keywords

        // we create a map directly from the given args array
        mp_map_t kw_args;
        mp_map_init_fixed_table(&kw_args, n_kw, args + n_args);

        // **** after this commit: actual call to native_base_init_wrapper(size_t n_args, const mp_obj_t *args, mp_map_t *kw_args)
        return self->fun.kw(n_args, args, &kw_args);

    } else {
        // function takes a variable number of arguments, but no keywords

        // **** before this commit: actual call to native_base_init_wrapper(size_t n_args, const mp_obj_t *args)
        return self->fun.var(n_args, args);
    }
}
```

- calling `super.__init__(*args)` or `super().__init__(**kwargs)` or combinations thereof i.e. anything with star arguments result in dispatching `MP_BC_CALL_METHOD_VAR_KW` which leads to the same call chain as above except that there's first an extra call to unpack the star arguments:

```
ENTRY(MP_BC_CALL_METHOD_VAR_KW):
 // We have following stack layout here:
 // fun self arg0 arg1 ... kw0 val0 kw1 val1 ... bitmap <- TOS
...

mp_obj_t mp_call_method_n_kw_var(bool have_self, size_t n_args_n_kw, const mp_obj_t *args) {
    mp_call_args_t out_args;
    mp_call_prepare_args_n_kw_var(have_self, n_args_n_kw, args, &out_args);
    mp_obj_t res = mp_call_function_n_kw(out_args.fun, out_args.n_args, out_args.n_kw, out_args.args);
}

...
```

So the main trade-off by this commit is that every call to the native base's make_new is going to be preceded by a call to `mp_map_init_fixed_table` even though most of what that does is unused and instead it merely serves as a way to pass the number of keyword arguments.

And the difference between `make_new` and 'normal' functions taking keyword arguments is that those are usually going to use the `kwargs` map directly and ignore that starting at `args + n_args` the keyword arguments also happen to be available.
Whereas make_new implementations which do use keyword arguments will usually repeat the
```
mp_map_t kwargs;
mp_map_init_fixed_table(&kwargs, n_kw, args + n_args)
```
part.

Reply:
Thanks for the detailed explanation.  You are right.

A different way to see the same thing is the following: if you use `MP_DEFINE_CONST_FUN_OBJ_KW` to wrap a C function, then that C function is always and only ever called from `fun_builtin_var_call`, along the path that has `self->sig & 1` true.  And since `fun_builtin_var_call` is passed the positional and kw arguments adjacently in an array, they are passed through to the wrapped C function.

So I think we can rely on this behaviour, and this PR is good.  But I suggest adding some comments:
- a comment just above the typedef for `mp_fun_kw_t` stating that, even though an `mp_map_t*` is passed, the `mp_obj_t*` array will actually contain the keyword arguments as well
- a comment in `fun_builtin_var_call` that `self->fun.kw` expects the `args` array to have both positional and keyword adjacent

(I guess one day we could change `mp_fun_kw_t`'s type so that it just takes the big `args` array and doesn't pass in a `mp_map_t`... but that's a big change.  Or could go the other way and make everything take a `mp_map_t`, but that's also a big change.)

Reply:
Ok thanks for checking, added comments, hope that's fine.

Reply:
That looks good, thanks!

> So the main trade-off by this commit is that every call to the native base's make_new is going to be preceded by a call to mp_map_init_fixed_table even though most of what that does is unused and instead it merely serves as a way to pass the number of keyword arguments.

Regarding this: yes, there will now be a slight additional overhead calling native base `make_new`, but IMO that's worth it for the added benefit of being able to pass in keyword arguments.  Constructors are anyway always going to be relatively slow due to memory allocation, and aren't as performance critical as normal functions/methods.